### PR TITLE
Bypass Calcite's SemiJoinRule and use our own.

### PR DESCRIPTION
**Pull request by gianm**
This simplifies DruidSemiJoin, which no longer needs to add aggregation back
in. It also allows some more kinds of queries to plan properly, like the one
added in "testTopNFilterJoin". I think this is because Calcite's builtin
SemiJoinRule didn't trigger when running on top of nodes that had already
been converted to DruidRels.

### CODE REVIEW

**Comment by fjy** [comment]
👍

**Comment by jihoonson** [comment]
I'm reviewing this.

**Comment by fjy** [comment]
@jihoonson any more comments?

**Comment by jihoonson** [comment]
@gianm @fjy, no it looks good to me, but conflicts need to be resolved.

File: sql/src/main/java/io/druid/sql/calcite/rule/DruidSemiJoinRule.java [hunk 1]
```diff
@@ -50,23 +99,66 @@ public static DruidSemiJoinRule create(final PlannerConfig plannerConfig)
...
+        );
+
+    if (bits.intersects(rightBits)) {
+      return;
+    }
+
+    final JoinInfo joinInfo = join.analyzeCondition();
+    final List<Integer> rightDimsOut = new ArrayList<>();
+    for (DimensionSpec dimensionSpec : right.getQueryBuilder().getGrouping().getDimensions()) {
+      rightDimsOut.add(right.getOutputRowSignature().getRowOrder().indexOf(dimensionSpec.getOutputName()));
```
Comment (gianm) [inline sql/src/main/java/io/druid/sql/calcite/rule/DruidSemiJoinRule.java]: Those are just field accesses so I think they're ok

**Comment by gianm** [comment]
Resolved conflicts

**Pull request merged**